File: cslox/cslox/Token.cs
namespace cslox
{
    internal class Token
	{
		internal enum TokenType
		{
			LEFT_PAREN, RIGHT_PAREN, LEFT_BRACE, RIGHT_BRACE,
			COMMA, DOT, MINUS, PLUS, SEMICOLON, SLASH, STAR,

			// One or two character tokens.
			BANG, BANG_EQUAL,
			EQUAL, EQUAL_EQUAL,
			GREATER, GREATER_EQUAL,
			LESS, LESS_EQUAL,

			// Literals.
			IDENTIFIER, STRING, NUMBER,

			// Keywords.
			AND, CLASS, ELSE, FALSE, FN, FOR, IF, NIL, OR,
			RETURN, SUPER, SELF, TRUE, VAR, WHILE, MUT, BREAK, CONTINUE,

			EOF,
			BUILTIN
		}

		internal static Dictionary<string, TokenType> Keywords = new Dictionary<string, TokenType>() {
			{ "and", TokenType.AND },
			{ "class", TokenType.CLASS },
			{ "else", TokenType.ELSE },
			{ "false", TokenType.FALSE },
			{ "fn", TokenType.FN },
			{ "for", TokenType.FOR },
			{ "if", TokenType.IF },
			{ "nil", TokenType.NIL },
			{ "or", TokenType.OR },
			{ "return", TokenType.RETURN },
			{ "super", TokenType.SUPER },
			{ "self", TokenType.SELF },
			{ "true", TokenType.TRUE },
			{ "var", TokenType.VAR },
			{ "while", TokenType.WHILE },
			{ "mut", TokenType.MUT },
			{ "break", TokenType.BREAK },
			{ "continue", TokenType.CONTINUE },
		};

		internal TokenType type;
		internal string lexeme;
		internal object? literal;
		internal int line;

		internal Token(TokenType type, string lexeme, object? literal, int line)
		{
			this.type = type;
			this.lexeme = lexeme;
			this.literal = literal;
			this.line = line;
		}

		public override string ToString()
		{
			return $"{type} {lexeme} {literal}";
		}
	}
}

File: hezen/hezen_runtime/src/token.rs
use std::fmt::Display;

use crate::error::HezenLineInfo;

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,

    // one or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    // literals.
    Identifier,
    String(String),
    Number(f64),

    // keywords.
    And,
    Class,
    Else,
    False,
    Fn,
    For,
    If,
    Nil,
    Or,
    Return,
    Super,
    Self_,
    True,
    Var,
    While,
    Mut,
    Break,
    Continue,

    Eof,
    Builtin,
}

impl Display for TokenType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenType::LeftParen => write!(f, "("),
            TokenType::RightParen => write!(f, ")"),
            TokenType::LeftBrace => write!(f, "{{"),
            TokenType::RightBrace => write!(f, "}}"),
            TokenType::Comma => write!(f, ","),
            TokenType::Dot => write!(f, "."),
            TokenType::Minus => write!(f, "-"),
            TokenType::Plus => write!(f, "+"),
            TokenType::Semicolon => write!(f, ";"),
            TokenType::Slash => write!(f, "/"),
            TokenType::Star => write!(f, "*"),
            TokenType::Bang => write!(f, "!"),
            TokenType::BangEqual => write!(f, "!="),
            TokenType::Equal => write!(f, "="),
            TokenType::EqualEqual => write!(f, "=="),
            TokenType::Greater => write!(f, ">"),
            TokenType::GreaterEqual => write!(f, ">="),
            TokenType::Less => write!(f, "<"),
            TokenType::LessEqual => write!(f, "<="),
            TokenType::Identifier => write!(f, "identifier"),
            TokenType::String(s) => write!(f, "\"{}\"", s),
            TokenType::Number(n) => write!(f, "{}", n),
            TokenType::And => write!(f, "and"),
            TokenType::Class => write!(f, "class"),
            TokenType::Else => write!(f, "else"),
            TokenType::False => write!(f, "false"),
            TokenType::Fn => write!(f, "fn"),
            TokenType::For => write!(f, "for"),
            TokenType::If => write!(f, "if"),
            TokenType::Nil => write!(f, "nil"),
            TokenType::Or => write!(f, "or"),
            TokenType::Return => write!(f, "return"),
            TokenType::Super => write!(f, "super"),
            TokenType::Self_ => write!(f, "self"),
            TokenType::True => write!(f, "true"),
            TokenType::Var => write!(f, "var"),
            TokenType::While => write!(f, "while"),
            TokenType::Mut => write!(f, "mut"),
            TokenType::Break => write!(f, "break"),
            TokenType::Continue => write!(f, "continue"),
            TokenType::Eof => write!(f, "eof"),
            TokenType::Builtin => write!(f, "builtin"),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Token {
    pub ty: TokenType,
    pub lexeme: String,
    pub position: HezenLineInfo,
}

impl Token {
    pub fn new(ty: TokenType, lexeme: String, position: HezenLineInfo) -> Self {
        Self {
            ty,
            lexeme,
            position,
        }
    }
}

impl Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.lexeme, self.ty)
    }
}

#[derive(Debug)]
pub struct Tokens {
    pub list: Vec<Token>,
}

impl Tokens {
    pub fn new() -> Self {
        Self { list: Vec::new() }
    }

    pub fn add(&mut self, token: Token) {
        self.list.push(token);
    }

    pub fn get(&self, index: usize) -> &Token {
        &self.list[index]
    }

    pub fn len(&self) -> usize {
        self.list.len()
    }
}

impl Display for Tokens {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for token in self.list.iter() {
            writeln!(f, "{}", token)?;
        }

        Ok(())
    }
}
